### ISSUE TITLE
`helm rollback` should not run patch logic as `--force` flag is not given (default behaviour)

### ISSUE THREAD

**Issue by liuming-dev** [helm/helm#8465]
Output of `helm version`: 
```
version.BuildInfo{Version:"v3.2.4", GitCommit:"0ad800ef43d3b826f31a5ad8dfbb4fe05d143688", GitTreeState:"dirty", GoVersion:"go1.14.3"}
```

Output of `kubectl version`: 
```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-27T00:38:11Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:48:36Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 

---

**Case**

`helm rollback` hits the patch logic and gives an error:

```
Rollback "emer-admin" failed: cannot patch "emer-admin" with kind Deployment: The order in patch list: [map[hostnames:[idea0101.mongodb] ip:10.153.44.147] map[hostnames:[noclickcpc0101.mongodb] ip:10.153.44.147] map[hostnames:[cpc0101.mongodb] ip:10.153.44.147] map[hostnames:[idea0102.mongodb] ip:10.153.52.224] map[hostnames:[noclickcpc0102.mongodb] ip:10.153.52.224] map[hostnames:[cpc0102.mongodb] ip:10.153.52.224] map[hostnames:[noclickcpc0103.mongodb] ip:10.153.53.139] map[hostnames:[idea0103.mongodb] ip:10.153.53.139] map[hostnames:[cpc0103.mongodb] ip:10.153.53.139] map[hostnames:[querykey0103.mongodb] ip:10.153.52.63] map[hostnames:[cpc0201.mongodb] ip:10.153.52.63] map[hostnames:[noclickcpc0201.mongodb] ip:10.153.52.63] map[hostnames:[idea0201.mongodb] ip:10.153.52.63] map[hostnames:[cpc0202.mongodb] ip:10.153.58.52] map[hostnames:[noclickcpc0202.mongodb] ip:10.153.58.52] map[hostnames:[idea0202.mongodb] ip:10.153.58.52] map[hostnames:[cpc0203.mongodb] ip:10.153.60.75] map[hostnames:[idea0203.mongodb] ip:10.153.60.75] map[hostnames:[noclickcpc0203.mongodb] ip:10.153.60.75] map[hostnames:[cpc0301.mongodb] ip:10.144.122.209] map[hostnames:[galaxywirelesslog01.mongodb] ip:10.144.122.209] map[hostnames:[noclickcpc0301.mongodb] ip:10.144.122.209] map[hostnames:[idea0301.mongodb] ip:10.144.122.209] map[hostnames:[galaxywirelesslog02.mongodb] ip:10.144.125.225] map[hostnames:[cpc0302.mongodb] ip:10.144.125.225] map[hostnames:[noclickcpc0302.mongodb] ip:10.144.125.225] map[hostnames:
```

**Analysis**

 As a default behaviour, `--force` flag is not set for `helm rollback`, which means `r.Force` value is `false`, and this will finally run `patch, patchType, err := createPatch(target, currentObj)`, diff among current release manifest, target release manifest and current manifest from k8s, then make a patch.

https://github.com/helm/helm/blob/e31818f2efd3e1e0867d9c189507c25c8102264a/pkg/action/rollback.go#L162

https://github.com/helm/helm/blob/e31818f2efd3e1e0867d9c189507c25c8102264a/pkg/kube/client.go#L204

https://github.com/helm/helm/blob/e31818f2efd3e1e0867d9c189507c25c8102264a/pkg/kube/client.go#L414-L453

I doubt this rationality, in my understanding, `rollback` means that going back to the given version, replacing the resources with totally the target release manifest rather than making a patch.

**Reproduce**

Here I provide an example to reproduce, to confirm that running patch logic may cause some unexpected result:

```sh
# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:41:55]
$ pwd
/Users/apple/.kube/helm-chart-sample

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:41:59]
$ ll
total 8
-rw-r--r--  1 apple  staff   1.1K  5  7 19:28 Chart.yaml
drwxr-xr-x  2 apple  staff    64B  5  7 19:28 charts
drwxr-xr-x  3 apple  staff    96B  7 16 11:40 templates
-rw-r--r--  1 apple  staff     0B  5  7 19:29 values.yaml

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:42:02]
$ ll templates
total 8
-rw-r--r--  1 apple  staff   569B  7 16 11:40 deployment.yaml

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:42:43]
$ cat templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: rollback-patch-deploy
spec:
  selector:
    matchLabels:
      paas.sogou.com/app: rollback-patch-deploy
  replicas: 1
  template:
    metadata:
      labels:
        paas.sogou.com/app: rollback-patch-deploy
    spec:
      containers:
      - name: busybox
        image: docker-reg.sogou-inc.com/official/busybox:1.31.1
      hostAliases:
      - hostnames:
        - liuming-dev.github.io
        ip: 127.0.0.1
      - hostnames:
        - liuming216448.sogou
        ip: 36.110.170.48
      - hostnames:
        - liuming.sogou
        ip: 36.110.170.48

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:42:48]
$ helm install rollback-patch-deploy ../helm-chart-sample
NAME: rollback-patch-deploy
LAST DEPLOYED: Thu Jul 16 11:43:00 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:43:00]
$ kubectl get deployment rollback-patch-deploy -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: rollback-patch-deploy
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2020-07-16T03:43:00Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  name: rollback-patch-deploy
  namespace: default
  resourceVersion: "2698527"
  selfLink: /apis/apps/v1/namespaces/default/deployments/rollback-patch-deploy
  uid: 5ceef022-ce82-438d-9a80-c27f5948702e
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      paas.sogou.com/app: rollback-patch-deploy
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        paas.sogou.com/app: rollback-patch-deploy
    spec:
      containers:
      - image: docker-reg.sogou-inc.com/official/busybox:1.31.1
        imagePullPolicy: IfNotPresent
        name: busybox
        resources: {}
  1 apiVersion: apps/v1
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      hostAliases:
      - hostnames:
        - liuming-dev.github.io
        ip: 127.0.0.1
      - hostnames:
        - liuming216448.sogou
        ip: 36.110.170.48
      - hostnames:
        - liuming.sogou
        ip: 36.110.170.48
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  conditions:
  - lastTransitionTime: "2020-07-16T03:43:00Z"
    lastUpdateTime: "2020-07-16T03:43:00Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  - lastTransitionTime: "2020-07-16T03:43:00Z"
    lastUpdateTime: "2020-07-16T03:43:00Z"
    message: ReplicaSet "rollback-patch-deploy-84c66b6ff" is progressing.
    reason: ReplicaSetUpdated
    status: "True"
    type: Progressing
  observedGeneration: 1
  replicas: 1
  unavailableReplicas: 1
  updatedReplicas: 1

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:43:04]
$ vi templates/deployment.yaml

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:43:42]
$ cat templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: rollback-patch-deploy
spec:
  selector:
    matchLabels:
      paas.sogou.com/app: rollback-patch-deploy
  replicas: 1
  template:
    metadata:
      labels:
        paas.sogou.com/app: rollback-patch-deploy
    spec:
      containers:
      - name: busybox
        image: docker-reg.sogou-inc.com/official/busybox:1.31.1
      hostAliases:
      - hostnames:
        - liuming-dev.github.io
        ip: 127.0.0.1
      - hostnames:
        - liuming216448.sogou
        - liuming.sogou
        ip: 36.110.170.48

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:43:46]
$ helm upgrade rollback-patch-deploy ../helm-chart-sample --force
Release "rollback-patch-deploy" has been upgraded. Happy Helming!
NAME: rollback-patch-deploy
LAST DEPLOYED: Thu Jul 16 11:43:55 2020
NAMESPACE: default
STATUS: deployed
REVISION: 2
TEST SUITE: None

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:43:55]
$ kubectl get deployment rollback-patch-deploy -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
    meta.helm.sh/release-name: rollback-patch-deploy
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2020-07-16T03:43:00Z"
  generation: 2
  labels:
    app.kubernetes.io/managed-by: Helm
  name: rollback-patch-deploy
  namespace: default
  resourceVersion: "2698684"
  selfLink: /apis/apps/v1/namespaces/default/deployments/rollback-patch-deploy
  uid: 5ceef022-ce82-438d-9a80-c27f5948702e
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      paas.sogou.com/app: rollback-patch-deploy
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        paas.sogou.com/app: rollback-patch-deploy
    spec:
      containers:
      - image: docker-reg.sogou-inc.com/official/busybox:1.31.1
        imagePullPolicy: IfNotPresent
        name: busybox
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      hostAliases:
      - hostnames:
        - liuming-dev.github.io
        ip: 127.0.0.1
      - hostnames:
        - liuming216448.sogou
        - liuming.sogou
        ip: 36.110.170.48
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  conditions:
  - lastTransitionTime: "2020-07-16T03:43:00Z"
    lastUpdateTime: "2020-07-16T03:43:00Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  - lastTransitionTime: "2020-07-16T03:43:00Z"
    lastUpdateTime: "2020-07-16T03:43:55Z"
    message: ReplicaSet "rollback-patch-deploy-5d8ddc8779" is progressing.
    reason: ReplicaSetUpdated
    status: "True"
    type: Progressing
  observedGeneration: 2
  replicas: 2
  unavailableReplicas: 2
  updatedReplicas: 1

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:44:02]
$ helm history rollback-patch-deploy
REVISION	UPDATED                 	STATUS    	CHART                  	APP VERSION	DESCRIPTION
1       	Thu Jul 16 11:43:00 2020	superseded	helm-chart-sample-0.1.0	1.16.0     	Install complete
2       	Thu Jul 16 11:43:55 2020	deployed  	helm-chart-sample-0.1.0	1.16.0     	Upgrade complete

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:44:10]
$ helm rollback rollback-patch-deploy 1
Rollback was a success! Happy Helming!

# apple @ liuming-dev in ~/.kube/helm-chart-sample [11:44:15]
$ kubectl get deployment rollback-patch-deploy -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "3"
    meta.helm.sh/release-name: rollback-patch-deploy
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2020-07-16T03:43:00Z"
  generation: 3
  labels:
    app.kubernetes.io/managed-by: Helm
  name: rollback-patch-deploy
  namespace: default
  resourceVersion: "2698768"
  selfLink: /apis/apps/v1/namespaces/default/deployments/rollback-patch-deploy
  uid: 5ceef022-ce82-438d-9a80-c27f5948702e
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      paas.sogou.com/app: rollback-patch-deploy
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        paas.sogou.com/app: rollback-patch-deploy
    spec:
      containers:
      - image: docker-reg.sogou-inc.com/official/busybox:1.31.1
        imagePullPolicy: IfNotPresent
        name: busybox
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      hostAliases:
      - hostnames:
        - liuming-dev.github.io
        ip: 127.0.0.1
      - hostnames:
        - liuming216448.sogou
        ip: 36.110.170.48
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  conditions:
  - lastTransitionTime: "2020-07-16T03:43:00Z"
    lastUpdateTime: "2020-07-16T03:43:00Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  - lastTransitionTime: "2020-07-16T03:43:00Z"
    lastUpdateTime: "2020-07-16T03:44:15Z"
    message: ReplicaSet "rollback-patch-deploy-666bff5685" is progressing.
    reason: ReplicaSetUpdated
    status: "True"
    type: Progressing
  observedGeneration: 3
  replicas: 2
  unavailableReplicas: 2
  updatedReplicas: 1

```

After rollback, it lost host `36.110.170.48 liuming.sogou`, this is not as expected. 

Similar topic issue: #7999 

**Todo**

- make `r.cfg.KubeClient.Update(current, target, true) `
https://github.com/helm/helm/blob/e31818f2efd3e1e0867d9c189507c25c8102264a/pkg/action/rollback.go#L162
- remove `--force` flag for `helm rollback`
-  add `--patch` flag if users want to do a patch

**Improvement**

Two ways to improve this case:

- remove `--force` flag, and add `--patch` flag, see PR:  #8473, #8474
- set `--force` default value to be true, see PR: #8475

**Comment by bacongobbler** [comment]
The omission of the force flag was entirely intentional from `helm rollback`. We documented this behaviour [in the FAQ](https://helm.sh/docs/faq/#improved-upgrade-strategy-3-way-strategic-merge-patches). Many service meshes and other controller-based applications inject data into Kubernetes objects. This can be something like a sidecar, labels, or other information. If we were to use the force flag, `helm rollback` would remove this information from the manifest. The goal of the rollback is to only roll back to the state that the user intended. Anything else injected by other means is preserved.

In other words, this is working as intended. It's unclear what we can do to fix your particular issue, but please stop writing PRs to "fix" things that have been documented as intentional behaviour.

**Comment by bacongobbler** [comment]
It looks like this case can be solved by using `helm rollback --force`. Please use that moving forward if you want to remove external metadata that is not managed by the chart.

**Comment by liuming-dev** [comment]
> It looks like this case can be solved by using `helm rollback --force`. Please use that moving forward if you want to remove external metadata that is not managed by the chart.

I have learned that this behaviour is designed intentionally, and set `--force` could resolve my case, thanks for  your suggestions @bacongobbler , and say sorry if this get you annoyed.
